File: expensereport-cxx/ExpenseReport.cpp
#include <chrono>
#include <sstream>
#include <iostream>
#include <iterator>
#include <algorithm>
#include <numeric>

#include "ExpenseReport.hpp"

void ExpenseReport::printReport(list<Expense> expenses, time_t now) {
    ostringstream result;
    header(now, result);
    body(expenses, result);
    summary(expenses, result);
    cout << result.str();
}

void ExpenseReport::body(const list<Expense> &expenses, ostringstream &result) const {
    for (auto &expense : expenses) {
        detail(expense, result);
    }
}

void ExpenseReport::detail(const Expense &expense, ostringstream &result) const {
    result << expense.getName() << '\t' << expense.amount << '\t' << expense.getOverLimitMarker() << '\n';
}

void ExpenseReport::summary(const list<Expense> &expenses, ostringstream &result) const {
    result << "Meal expenses: " << sumMeals(expenses) << '\n';
    result << "Total expenses: " << sumTotal(expenses) << '\n';
}

int ExpenseReport::sumTotal(const list<Expense> &expenses) const {
    return accumulate(begin(expenses), end(expenses), 0, [](int total, const Expense &expense) {
        return total + expense.amount;
    });
}

int ExpenseReport::sumMeals(const list<Expense> &expenses) const {
    return accumulate(begin(expenses), end(expenses), 0, [](int total, const Expense &expense) {
        if (expense.isMeal()) { return total + expense.amount; }
        return total;
    });
}

void ExpenseReport::header(time_t &now, ostringstream &result) const {
    result << "Expenses " << ctime(&now) << '\n';
}

string Expense::getName() const {
    return type.name;
}

string Expense::getOverLimitMarker() const {
    return isOverLimit() ? "X" : " ";
}

bool Expense::isOverLimit() const { return amount > type.limit; }

bool Expense::isMeal() const {
    return type.isMeal;
}

Expense::Expense(ExpenseType type, int amount) : type(type), amount(amount) {
}

ExpenseType::ExpenseType(const string &name, int limit, bool isMeal) :
        name(name), limit(limit), isMeal(isMeal) {
}

File: expensereport-php/src/ExpenseReport.php
<?php

namespace ExpenseReport;

class ExpenseReport
{

    function print_report($expenses, $timestamp)
    {
        $this->header($timestamp);
        $this->body($expenses);
        $this->summary($expenses);
    }

    public function header($timestamp): void
    {
        $date = date("Y-m-d h:i:sa", $timestamp);
        print("Expense Report {$date}\n");
    }

    public function body($expenses)
    {
        foreach ($expenses as $expense) {
            $this->detail($expense);
        }
    }

    public function summary($expenses): void
    {
        print("Meal Expenses: " . $this->sumMeals($expenses) . "\n");
        print("Total Expenses: " . $this->sumTotal($expenses) . "\n");
    }

    public function detail($expense): void
    {
        print($expense->getExpenseName() . "\t" . $expense->amount . "\t" . $this->getOverExpensesMarker($expense) . "\n");
    }

    public function sumTotal($expenses)
    {
        $total = 0;
        foreach ($expenses as $expense) {
            $total += $expense->amount;
        }
        return $total;
    }

    public function sumMeals($expenses)
    {
        return $this->sumTotal(array_filter($expenses, function ($expense) {
            return $expense->isMeal();
        }));
    }

    public function getOverExpensesMarker($expense): string
    {
        $mealOverExpensesMarker = $expense->isOverLimit() ? "X" : " ";
        return $mealOverExpensesMarker;
    }

}
